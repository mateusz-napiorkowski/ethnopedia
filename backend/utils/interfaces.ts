export interface artworkCategory {
    name: string
    value: string
    subcategories: Array<artworkCategory>
}

export interface record {
    categories: Array<artworkCategory>,
    collectionName: string
}

export interface collectionCategory {
    name: string,
    subcategories: Array<collectionCategory>
}

export interface fileToDelete {
    originalFilename: string,
    filePath: string,
    size: number,
    uploadedAt: Date,
    _id: string
}