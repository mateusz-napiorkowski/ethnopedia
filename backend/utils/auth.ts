import { Request } from "express"
const jwt = require("jsonwebtoken")

export const checkUserIsLoggedIn = async (req: Request) => {
    try {
        const token = req.headers.authorization?.split(" ")[1]
        if (!token) return false;
        try {
            const decoded = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET as string)
            return true
        } catch (error) {
            return false
        }
    } catch (error) {
        return false
    }
}

module.exports = { checkUserIsLoggedIn }