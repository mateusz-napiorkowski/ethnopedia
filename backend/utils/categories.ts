import CollectionCollection from "../models/collection";
import { collectionCategory, artworkCategory } from "./interfaces"

export const hasValidCategoryFormat = (categories: Array<collectionCategory>, isRootArray = true): boolean => {
    if(isRootArray && !categories.length)
        return false
    return categories.every(category => 
        category?.name && 
        Array.isArray(category?.subcategories) && 
        hasValidCategoryFormat(category.subcategories, false)
    );
};

export const artworkCategoriesHaveValidFormat = (artworkCategories: Array<artworkCategory>, collectionCategories: Array<collectionCategory>, isRootArray = true): boolean => {
    if (isRootArray && (artworkCategories.length == 0 || collectionCategories.length == 0)) return false;
    if (artworkCategories.length !== collectionCategories.length) return false;

    return collectionCategories.every((category, index) => {
        const artworkCategory = artworkCategories[index];

        return (
            category.name === artworkCategory.name &&
            !!artworkCategory.value &&
            artworkCategoriesHaveValidFormat(artworkCategory.subcategories, category.subcategories, false)
        );
    });
};

export const isValidCollectionCategoryStructureForCollectionUpdate = (artworkSubcategories: Array<artworkCategory>, collectionSubcategories: Array<collectionCategory>): boolean => {
    if(artworkSubcategories.length > collectionSubcategories.length) return false
    return artworkSubcategories.every((artworkCategory, index) =>
        isValidCollectionCategoryStructureForCollectionUpdate(
          artworkCategory.subcategories,
          collectionSubcategories[index].subcategories
        )
      );
}

const getNestedCategories = ((prefix: string, subcategories: Array<artworkCategory>) => {
    const nestedCategories: Array<string> = []
    for(const subcategory of subcategories) {
        nestedCategories.push(`${prefix}.${subcategory.name}`)
        nestedCategories.push(...getNestedCategories(`${prefix}.${subcategory.name}`, subcategory.subcategories))
    }
    return nestedCategories
})

export const getAllCategories = async (collectionName: string) => {
    try {
        const collections = await CollectionCollection.find({name: collectionName}).exec()
        if(collections.length !== 1)
            throw new Error("Collection not found")
        const allCategories: Array<string> = []
        for(const category of collections[0].categories) {
            allCategories.push(category.name)
            allCategories.push(...getNestedCategories(`${category.name}`, category.subcategories))
        }           
        return allCategories
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Collection not found`)
            throw new Error("Collection not found")
        else
            throw new Error("Database unavailable")
    }
}

export const transformCategoriesArrayToCategoriesObject = (categoriesArray: Array<string>, depth = 1) => {
    const categories: Array<any> = []
    categoriesArray.forEach(field => {
        const allSubcategoriesNames = categoriesArray
            .filter(columnName => (columnName.startsWith(`${field}.`)))
            .map(item => item.split('.').slice(1).join('.'));
        if(!field.includes("."))
            categories.push({
                name: field,
                subcategories: transformCategoriesArrayToCategoriesObject(allSubcategoriesNames, depth + 1)
            })        
    })
    return categories
}

export const findMissingParentCategories = ((categoriesArray: Array<string>) => {
    const missingCategories: Array<string> = []
    categoriesArray.forEach((category: string) => {
        const parentCategory = category.split(".").slice(0, -1).join(".")
        if(category.includes(".") && !categoriesArray.includes(parentCategory))
            missingCategories.push(parentCategory)
    });
    return missingCategories
})