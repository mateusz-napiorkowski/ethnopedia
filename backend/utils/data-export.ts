import { subcategoryData } from "./interfaces"

const findValue: any = (subcategories: Array<subcategoryData>, categoryNameSplitByDot: Array<string>) => {
    const categoryDepth = categoryNameSplitByDot.length
    const [topmostParentCategoryName, ...categoryNameWithoutTopmostPart] = categoryNameSplitByDot;
    if(categoryDepth > 1) {
        const matchingSubcategory = subcategories.find(subcategory => subcategory.name === topmostParentCategoryName);
        return matchingSubcategory ? findValue(matchingSubcategory.subcategories, categoryNameWithoutTopmostPart) : '';
    }
    const matchingCategory = subcategories.find(category => category.name === topmostParentCategoryName);
    return matchingCategory ? matchingCategory.values.join(';') : '';
}

export const fillRow = (keys: Array<string>, categories: Array<subcategoryData>) => {
    const rowdata: any = {}
    keys.forEach(key => {
        rowdata[key] = findValue(categories, key.split("."))
    });
    return rowdata
}