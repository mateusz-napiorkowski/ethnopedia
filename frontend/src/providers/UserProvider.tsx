import React, { createContext, useContext, useState, useEffect } from "react"
import { jwtDecode } from "jwt-decode"

export interface JWT {
    exp: number;
    username: string;
    firstName: string;
    userId: string;
}

interface UserContextProps {
    isUserLoggedIn: boolean;
    firstName: string;
    userId: string;
    jwtToken: string;
    setUserData: (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => void;
}

export const UserContext = createContext<UserContextProps>({
    isUserLoggedIn: false,
    firstName: "",
    userId: "",
    jwtToken: "",
    setUserData: () => {
    },
})

export const useUser = () => useContext(UserContext)

interface UserProviderProps {
    children: React.ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [isUserLoggedIn, setIsUserLoggedIn] = useState(false)
    const [firstName, setFirstName] = useState("")
    const [userId, setUserId] = useState("")
    const [jwtToken, setJwtToken] = useState("")

    const setUserData = (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => {
        setIsUserLoggedIn(isLoggedIn)
        setFirstName(name)
        setUserId(userId)
        setJwtToken(jwtToken)
    }

    useEffect(() => {
        const token = localStorage.getItem("token")

        if (token) {
            try {
                const decodedToken = jwtDecode<JWT>(token)
                const currentTime = Date.now() / 1000
                if (decodedToken.exp > currentTime) {
                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                } else {
                    localStorage.removeItem("token")
                    setUserData(false, "", "", "")
                }
            } catch (error) {
                console.error("Invalid token:", error)
                localStorage.removeItem("token")
                setUserData(false, "", "", "")
            }
        }
    }, [])

    return (
        <UserContext.Provider value={{ isUserLoggedIn, firstName, userId, jwtToken, setUserData }}>
            {children}
        </UserContext.Provider>
    )
}
