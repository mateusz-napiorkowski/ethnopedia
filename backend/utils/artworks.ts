import { getAllCategories } from "./categories";

const fillSubcategoriesFilterPart: any = (searchText: string, currentDepth: number, maxDepth: number) => {
    if (maxDepth === 0) return []
    return {
        $elemMatch: {
            $or: currentDepth === maxDepth 
                ? [{ values: [searchText] }] 
                : [
                    { values: [searchText] },
                    { subcategories: fillSubcategoriesFilterPart(searchText, currentDepth + 1, maxDepth) }
                ]
        }
    }
}

export const constructQuickSearchFilter = async (searchText: any, collectionName: string) => {
    const allCategories = await getAllCategories(collectionName)
    const maxDepth = allCategories.length > 0 ?
        Math.max.apply(Math, allCategories.map((category) => category.split('.').length)) : 0
    const queryFilter = {
        collectionName: collectionName,
        categories: fillSubcategoriesFilterPart(searchText, 1, maxDepth)
    }
    return queryFilter
}

const constructAdvSearchSubcategoriesFilter = (searchRules: Array<Array<string>>, depth: number) => {
    let subcategoryFilter: any = {$all: []}
    searchRules.forEach(([subcategoryName, subcategoryValue]) => {
        const subcategoryNameSplitByDot = subcategoryName.split('.');
        const isCurrentDepthSubcategory = subcategoryNameSplitByDot.length === depth;
        if (!isCurrentDepthSubcategory) return

        const deeperSubcategoriesSearchRules = searchRules.filter(([deeperSubcategoryName]) => 
            deeperSubcategoryName.startsWith(`${subcategoryName}.`)
        );

        let newFilterPart: any = {
            $elemMatch: {
                name: subcategoryNameSplitByDot.slice(depth - 1).join('.'),
            }
        };

        if(subcategoryValue)
            newFilterPart.$elemMatch.values = [subcategoryValue]

        if(deeperSubcategoriesSearchRules.length != 0)
            newFilterPart.$elemMatch.subcategories = constructAdvSearchSubcategoriesFilter(deeperSubcategoriesSearchRules, depth + 1)

        subcategoryFilter.$all.push(newFilterPart)
    })
    return subcategoryFilter
}

const getOnlySearchRulesArray = (reqQuery: any) => {
    let rulesArray: any = []
    for(const categoryName in reqQuery) {
        if(!["page", "pageSize", "sortOrder", "search"].includes(categoryName)) {
            rulesArray.push([categoryName, reqQuery[categoryName]])
            const categoryNameSplitByDot = categoryName.split('.')
            while(categoryNameSplitByDot.length !== 1) {
                categoryNameSplitByDot.pop()
                if(!reqQuery.hasOwnProperty(categoryNameSplitByDot.join('.'))) {
                    rulesArray.push([categoryNameSplitByDot.join('.'), undefined])
                }
            }
        }
    }
    return rulesArray
}

export const constructAdvSearchFilter = (requestQuery: any, collectionName: string) => {
    const searchRules = getOnlySearchRulesArray(requestQuery)

    const queryFilter: any = {
        collectionName: collectionName,
        categories: {$all: []}
    }

    searchRules.forEach(([categoryName, categoryValue]: [string, any]) => {
        if(categoryName.includes('.')) return //skip subcategory search rules

        const currentCategorySubcategoriesSearchRules = searchRules.filter(([subcategoryName]: [string, any]) => 
            subcategoryName.startsWith(`${categoryName}.`)
        );

        const categoryFilter: any = { 
            $elemMatch: { 
                name: categoryName 
            }
        };

        if(categoryValue)
            categoryFilter.$elemMatch.values = [categoryValue];
        
        if(currentCategorySubcategoriesSearchRules.length > 0)
            categoryFilter.$elemMatch.subcategories = constructAdvSearchSubcategoriesFilter(currentCategorySubcategoriesSearchRules, 2);

        queryFilter.categories.$all.push(categoryFilter);
    })
    return queryFilter
}

const findMatchingCategory = (categoryName: string, record: any) => {
    if(categoryName.split(".").length === 1) {
        if(!record.subcategories)
            return record.categories.find((category: any) => category.name === categoryName);
        return record.subcategories.find((category: any) => category.name === categoryName);
    } else {
        const prefix = categoryName.split(".")[0]
        const rest = categoryName.split(".").slice(1).join(".")
        if(!record.subcategories) {
            const parent = record.categories.find((category: any) => category.name === prefix);
            if(!parent)
                return undefined
            return findMatchingCategory(rest, parent)
        }
        const parent = record.subcategories.find((category: any) => category.name === prefix);
        if(!parent)
            return undefined
        return findMatchingCategory(rest, parent)
    }
}

export const sortRecordsByCategory = (records: any, order: string) => {
    const [categoryToSortBy, ascOrDesc] = order.split('-')

    const recordAndCategoryValuePairs = records.map((record: any) => {
        const matchingCategory = findMatchingCategory(categoryToSortBy, record)
        const categoryValue = matchingCategory ? matchingCategory.values.join(", ") : null;
        return [record, categoryValue];
    })
    .sort((a: any, b: any) => {
        if (a[1] && b[1])
            return a[1].toUpperCase().localeCompare(b[1].toUpperCase())
        
        //records which have the category to sort by come first
        if (a[1]) return -1;
        if (b[1]) return 1;
    });    
    
    const sortedRecords = recordAndCategoryValuePairs.map((pair: any) => pair[0]);
    if(ascOrDesc == "desc")
        return sortedRecords.reverse()
    return sortedRecords
}