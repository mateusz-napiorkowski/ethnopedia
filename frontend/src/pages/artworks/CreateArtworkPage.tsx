import React, { useEffect, useState, useMemo } from 'react';
import { useQuery, useQueryClient } from 'react-query';
import { Formik, Form, FormikHelpers } from 'formik';
import { useNavigate, useParams } from 'react-router-dom';
import Navbar from '../../components/navbar/Navbar';
import Navigation from '../../components/Navigation';
import LoadingPage from '../LoadingPage';
import { useUser } from '../../providers/UserProvider';
import { getAllCategories } from '../../api/categories';
import { getCollection } from '../../api/collections';
import { getArtwork, createArtwork, editArtwork } from '../../api/artworks';
import { getArtworksForPage } from '../../api/artworks'; // Import for getting all artworks
import MetadataForm from '../../components/artwork/MetadataForm';
import { Metadata } from '../../@types/Metadata';

interface FormValues {
    categories: Metadata[];
}

const CreateArtworkPage: React.FC = () => {
    const { collectionId, artworkId } = useParams<{
        collectionId: string;
        artworkId?: string;
    }>();
    const queryClient = useQueryClient();
    const navigate = useNavigate();
    const { jwtToken } = useUser();

    // Pobranie kategorii (w formacie dot.notation) lub metadanych rekordu (w trybie edycji)
    const { data: catData, isLoading: catsLoading } = useQuery(
        ['categories', collectionId],
        () => getAllCategories([collectionId as string]),
        { enabled: !!collectionId }
    );

    const { data: collData } = useQuery(
        ['collection', collectionId],
        () => getCollection(collectionId!),
        { enabled: !!collectionId }
    );



    // Fetch all artworks for suggestions
    const { data: artworksData } = useQuery(
        ['allArtworks', collectionId],
        () => getArtworksForPage(
            [collectionId as string],
            1,
            1000,
            'createdAt-desc',
            '',
            {}
        ),
        { enabled: !!collectionId }
    );

    const [initialCategoryPaths, setInitialCategoryPaths] = useState<string[]>([]);
    const [initialMetadataTree, setInitialMetadataTree] = useState<Metadata[] | undefined>(
        undefined
    );

    // Extract suggestions from existing artworks
    const suggestionsByCategory = useMemo(() => {
        const suggestions: Record<string, Set<string>> = {};

        // Function to extract values from artwork categories
        const extractValues = (categories: Metadata[], prefix: string = '') => {
            categories.forEach((category) => {
                const categoryPath = prefix ? `${prefix}.${category.name}` : category.name;

                if (category.value && category.value.trim()) {
                    if (!suggestions[categoryPath]) {
                        suggestions[categoryPath] = new Set();
                    }
                    suggestions[categoryPath].add(category.value.trim());
                }

                if (category.subcategories && category.subcategories.length > 0) {
                    extractValues(category.subcategories, categoryPath);
                }
            });
        };

        // Extract from all artworks
        artworksData.artworks.forEach((artwork: any, index: number) => {
            if (artwork.categories) {
                extractValues(artwork.categories);
            }
        });

        // Convert Sets to Arrays and sort
        const result: Record<string, string[]> = {};
        Object.keys(suggestions).forEach(key => {
            result[key] = Array.from(suggestions[key]).sort();
        });

        return result;
    }, [artworksData]);

    useEffect(() => {
        if (artworkId) {
            getArtwork(artworkId).then((res) => {
                setInitialMetadataTree(res.artwork.categories);
            });
        } else if (catData?.categories) {
            setInitialCategoryPaths(catData.categories);
        }
    }, [artworkId, catData]);

    if (catsLoading || (!initialCategoryPaths.length && !artworkId)) {
        return <LoadingPage />;
    }

    return (
        <div className="min-h-screen flex flex-col overflow-y-auto">
            <Navbar />
            <div className="container px-8 mt-6 max-w-3xl mx-auto">
                <Navigation />
                <h2 className="text-2xl font-bold mt-2">
                    {artworkId ? 'Edytuj rekord z kolekcji' : 'Dodaj nowy rekord do kolekcji'}
                </h2>
                <h2 className="text-2xl mb-4">
                    {collData?.name}
                </h2>

                <Formik<FormValues>
                    initialValues={{categories: initialMetadataTree || []}}
                    enableReinitialize
                    validate={(values) => {
                        const errs: Partial<Record<keyof FormValues, string>> = {};
                        const anyFilled = values.categories.some(
                            (c) => (c.value ?? '').trim().length > 0
                        );
                        if (!anyFilled) {
                            errs.categories = 'Przynajmniej jedno pole musi być wypełnione.';
                        }
                        return errs;
                    }}
                    onSubmit={async (
                        values,
                        {setSubmitting}: FormikHelpers<FormValues>
                    ) => {
                        const payload = {
                            categories: values.categories,
                            collectionName: collData?.name,
                        };
                        try {
                            if (artworkId) {
                                await editArtwork(payload, artworkId, jwtToken!);
                            } else {
                                await createArtwork(payload, jwtToken!);
                            }
                            queryClient.invalidateQueries(['artworks', collectionId]);
                            queryClient.invalidateQueries(['allArtworks', collectionId]); // Invalidate suggestions cache
                            navigate(-1);
                        } catch (e) {
                            console.error(e);
                        } finally {
                            setSubmitting(false);
                        }
                    }}
                >
                    {({setFieldValue, isSubmitting, errors, touched}) => (
                        <Form>
                            <MetadataForm
                                initialMetadataTree={initialMetadataTree}
                                categoryPaths={initialCategoryPaths}
                                setFieldValue={setFieldValue}
                                suggestionsByCategory={suggestionsByCategory}
                            />
                            {/* globalny komunikat jeśli walidacja nie przeszła */}
                            {typeof errors.categories === 'string' && touched.categories && (
                                <p className="mt-2 text-red-500">{errors.categories}</p>
                            )}
                            {/* sticky bottom bar */}
                            <div
                                className="fixed bottom-0 left-0 w-full bg-white dark:bg-gray-900 border-t border-gray-300 dark:border-gray-500 py-3 px-4 flex justify-end gap-2 shadow-[0_-2px_4px_rgba(0,0,0,0.05)] z-50">

                                <div className="max-w-3xl w-full mx-auto flex justify-end">
                                    <button
                                        type="button"
                                        onClick={() => navigate(-1)}
                                        className="px-4 py-2 border rounded"
                                    >
                                        Anuluj
                                    </button>
                                    <button
                                        type="submit"
                                        disabled={isSubmitting}
                                        className="px-4 py-2 bg-blue-600 text-white rounded ml-2"
                                    >
                                        {artworkId ? 'Zapisz' : 'Utwórz'}
                                    </button>
                                </div>
                            </div>

                            <div className="h-20"/>
                            {/* Przestrzeń pod sticky barem */}

                        </Form>
                    )}
                </Formik>
            </div>
        </div>
    );
};

export default CreateArtworkPage;