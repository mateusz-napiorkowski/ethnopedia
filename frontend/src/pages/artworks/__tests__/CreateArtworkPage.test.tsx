import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import CreateArtworkPage from "../CreateArtworkPage"
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { UserContext } from '../../../providers/UserProvider';
import {jwtToken, collectionData, artworkIds} from './utils/consts'

const mockGetArtwork = jest.fn();
const mockGetArtworksForPage = jest.fn()
const mockCreateArtwork = jest.fn();
const mockEditArtwork = jest.fn();

jest.mock('../../../api/artworks', () => ({
  __esModule: true,
  getArtwork: () => mockGetArtwork(),
  getArtworksForPage: () => mockGetArtworksForPage(),
  createArtwork: (collectionId: string, payload: any, filesToUpload: File[], jwtToken: string) => mockCreateArtwork(collectionId, payload, filesToUpload, jwtToken),
  editArtwork: () => mockEditArtwork()
}));

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate,
}));

const mockGetAllCategories = jest.fn()
jest.mock('../../../api/categories', () => ({
    getAllCategories: () => mockGetAllCategories(),
}))

const mockGetCollection = jest.fn()
jest.mock('../../../api/collections', () => ({
    getCollection: () => mockGetCollection
}))

const queryClient = new QueryClient();
const user = userEvent.setup()

const UserContextProps = {
    isUserLoggedIn: true,
    firstName: "123",
    userId: "66b6506fbb64df165e8a9ce6",
    jwtToken: jwtToken,
    setUserData: jest.fn()
}

const renderPage = (edit = false) => {
        return render(
            <UserContext.Provider value={ UserContextProps }>
                <QueryClientProvider client={queryClient}>
                    <MemoryRouter initialEntries={[
                        edit ? 
                            `/collections/${collectionData._id}/artworks/${artworkIds[0]}/edit-artwork`
                            : `/collections/${collectionData._id}/create-artwork/`
                        ]}>
                        <Routes>
                            <Route path={
                                edit ?
                                    "/collections/:collectionId/artworks/:artworkId/edit-artwork"
                                    : "/collections/:collectionId/create-artwork/"
                            } element={<CreateArtworkPage/>}/>
                        </Routes>  
                    </MemoryRouter>
                </QueryClientProvider>
            </UserContext.Provider>
            
        );
};

describe("CreateArtworkPage tests", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        queryClient.clear();
    });

    it("should render loading state", () => {
        const {getByTestId, queryByTestId} = renderPage()

        expect(getByTestId('loading-page-container')).toBeInTheDocument()
        expect(queryByTestId('create-artwork-page-container')).not.toBeInTheDocument()
    })

    it("should render component after data is fetched from API and render category tree correctly", async () => {
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValue({categories: ["Tytuł", "Tytuł.Podtytuł", "Artyści", "Rok"]})
        mockGetArtworksForPage.mockReturnValue({artworks: [], total: 0, currentPage: 1, pageSize: 1000})
        const {getByTestId, queryByTestId} = renderPage()
        await waitFor(() => getByTestId('create-artwork-page-container'))

        expect(getByTestId('create-artwork-page-container')).toBeInTheDocument()
        expect(queryByTestId('loading-page-container')).not.toBeInTheDocument()
        expect(getByTestId("category-tree")).toMatchSnapshot()
    })

    it("should show error message when trying to create artwork with none of the category input fields filled in", async () => {
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValue({categories: ["Tytuł", "Tytuł.Podtytuł", "Artyści", "Rok"]})
        mockGetArtworksForPage.mockReturnValue({artworks: [], total: 0, currentPage: 1, pageSize: 1000})
        const {getByTestId, getByText} = renderPage()
        await waitFor(() => getByTestId('create-artwork-page-container'))
        const createArtworkButton = getByText(/utwórz/i)

        await user.click(createArtworkButton)
        expect(getByText(/przynajmniej jedno pole musi być wypełnione./i)).toBeInTheDocument()
    })

    it("should call createArtwork when create artwork is clicked and data provided in the form is correct", async () => {
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValue({categories: ["Tytuł", "Tytuł.Podtytuł", "Artyści", "Rok"]})
        mockGetArtworksForPage.mockReturnValue({artworks: [], total: 0, currentPage: 1, pageSize: 1000})
        mockCreateArtwork.mockReturnValue({
            "artwork": {
                "collectionName": collectionData.name,
                "categories": [
                    {
                        "name": "Tytuł",
                        "value": "an artwork title",
                        "subcategories": [
                            {
                                "name": "Podtytuł",
                                "value": "",
                                "subcategories": []
                            }
                        ]
                    },
                    {
                        "name": "Artyści",
                        "value": "",
                        "subcategories": []
                    },
                    {
                        "name": "Rok",
                        "value": "",
                        "subcategories": []
                    }
                ],
                "_id": "68fd30ad0db5373b10dd000d",
                "files": [],
                "createdAt": "2025-10-25T20:18:53.913Z",
                "updatedAt": "2025-10-25T20:18:53.913Z",
                "__v": 0
            },
            "uploadedFilesCount": 0,
            "failedUploadsCount": 0,
            "failedUploadsCauses": []
        })
        const {getByTestId, getByText, getByLabelText} = renderPage()
        await waitFor(() => getByTestId('create-artwork-page-container'))
        const createArtworkButton = getByText(/utwórz/i)
        const titleInputField = getByLabelText("Tytuł-input")

        await user.type(titleInputField, "an artwork title")
        await user.click(createArtworkButton)

        expect(mockCreateArtwork).toHaveBeenCalledWith(
            collectionData._id,
            [
                {
                    "name": "Tytuł",
                    "value": "an artwork title",
                    "subcategories": [
                        {
                            "name": "Podtytuł",
                            "value": "",
                            "subcategories": []
                        }
                    ]
                },
                {
                    "name": "Artyści",
                    "value": "",
                    "subcategories": []
                },
                {
                    "name": "Rok",
                    "value": "",
                    "subcategories": []
                }
            ],
            [],
            UserContextProps.jwtToken
        )
        expect(mockUseNavigate).toHaveBeenCalledWith(-1)
    })

    it("should call useNavigate(-1) when cancel button is clicked", async () => {
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValue({categories: ["Tytuł", "Tytuł.Podtytuł", "Artyści", "Rok"]})
        mockGetArtworksForPage.mockReturnValue({artworks: [], total: 0, currentPage: 1, pageSize: 1000})
        const {getByTestId, getByText} = renderPage()
        await waitFor(() => getByTestId('create-artwork-page-container'))
        const cancelButton = getByText(/anuluj/i)

        await user.click(cancelButton)
        expect(mockUseNavigate).toHaveBeenCalledWith(-1)
    })
})