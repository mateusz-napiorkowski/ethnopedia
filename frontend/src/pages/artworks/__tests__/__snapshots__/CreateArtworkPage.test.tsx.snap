// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateArtworkPage tests should render component after data is fetched from API and render category tree correctly 1`] = `
<div
  data-testid="category-tree"
>
  <div
    class="relative "
  >
    <div
      class="relative w-full mb-2"
    >
      <label
        class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800"
        for="field-0"
      >
        Tytuł
      </label>
      <input
        aria-label="Tytuł-input"
        class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
        id="field-0"
        maxlength="100"
        type="text"
        value=""
      />
    </div>
    <div
      class="ml-4"
    >
      <div
        class="relative pl-6"
      >
        <div
          class="absolute left-0 w-px bg-gray-300 dark:bg-gray-600"
          style="top: -0.5rem; bottom: 2.1rem;"
        />
        <div
          class="absolute left-0 top-6 h-px w-6 bg-gray-300 dark:bg-gray-600"
        />
        <div
          class="relative w-full mb-2"
        >
          <label
            class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800"
            for="field-0-0"
          >
            Podtytuł
          </label>
          <input
            aria-label="Podtytuł-input"
            class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
            id="field-0-0"
            maxlength="100"
            type="text"
            value=""
          />
        </div>
        <div
          class="ml-4"
        />
      </div>
    </div>
  </div>
  <div
    class="relative "
  >
    <div
      class="relative w-full mb-2"
    >
      <label
        class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800"
        for="field-1"
      >
        Artyści
      </label>
      <input
        aria-label="Artyści-input"
        class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
        id="field-1"
        maxlength="100"
        type="text"
        value=""
      />
    </div>
    <div
      class="ml-4"
    />
  </div>
  <div
    class="relative "
  >
    <div
      class="relative w-full mb-2"
    >
      <label
        class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800"
        for="field-2"
      >
        Rok
      </label>
      <input
        aria-label="Rok-input"
        class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
        id="field-2"
        maxlength="100"
        type="text"
        value=""
      />
    </div>
    <div
      class="ml-4"
    />
  </div>
</div>
`;
